### PR TITLE
Permit to retrieve raw bibentry for each bibtex entries

### PR DESCRIPTION
**Pull request by flopraden**
- Adding a RAW key in the bibtex entry like ID extra key
- I didn't add a dictionary for raw text to stick with the way of ID/ENTRYTYPE

### CODE REVIEW

**Comment by sciunto** [comment]
I don't think we want this feature activated by default. We need a boolean to activate that :) + a unittest.

**Comment by flopraden** [comment]
I'll write it :) and get some time on unittest

**Comment by flopraden** [comment]
I don't know if the unittest is as you expected but it should test the feature. I corrected also the bwriter (thx to unittest)

File: bibtexparser/bibtexexpression.py [hunk 1]
```diff
@@ -65,6 +65,18 @@ def in_braces_or_pars(exp):
     return ((pp.Suppress('{') + exp + pp.Suppress('}')) |
             (pp.Suppress('(') + exp + pp.Suppress(')')))
 
+# add raw matched text at begin of all token
+def rawExtraction(expr):
```
Comment (sciunto) [inline bibtexparser/bibtexexpression.py]: Could you had a docstring here please?

File: bibtexparser/bparser.py [hunk 2]
```diff
@@ -65,7 +65,8 @@ def __init__(self, data=None,
                  customization=None,
                  ignore_nonstandard_types=True,
                  homogenize_fields=False,
-                 common_strings=False):
+                 common_strings=False,
+                 add_raw_bibtex_entry=False):
```
Comment (sciunto) [inline bibtexparser/bparser.py]: This new option needs to be documented. :)

File: bibtexparser/bparser.py [hunk 3]
```diff
@@ -99,6 +100,9 @@ def __init__(self, data=None,
         # hangs We are going to default to utf8, and mandate it.
         self.encoding = 'utf8'
 
+        # Add RAW element in bibentry corresponding to the raw bibtex entry as string
+        self.add_raw_bibtex_entry=add_raw_bibtex_entry
```
Comment (sciunto) [inline bibtexparser/bparser.py]: Spaces around = (pep8)

File: bibtexparser/bibtexexpression.py [hunk 4]
```diff
@@ -65,6 +65,18 @@ def in_braces_or_pars(exp):
...
+    locMarker = pp.Empty().setParseAction(lambda s,loc,t: loc)
+    endlocMarker = locMarker.copy()
+    endlocMarker.callPreparse = False
+    matchExpr = locMarker("_original_start") + expr + endlocMarker("_original_end")
+    def addRawText(s,l,t):
```
Comment (sciunto) [inline bibtexparser/bibtexexpression.py]: spaces after comas

File: bibtexparser/bibtexexpression.py [hunk 5]
```diff
@@ -156,10 +168,15 @@ def __init__(self):
...
 
         # Entry: type, key, and fields
-        self.entry = (entry_type +
-                      in_braces_or_pars(key + pp.Suppress(',') + field_list)
-                      )('Entry')
-
+        if add_raw_bibtex_entry:
+            self.entry = (rawExtraction(entry_type +
+                                        in_braces_or_pars(key + pp.Suppress(',') + field_list)
+            ))('Entry')
```
Comment (sciunto) [inline bibtexparser/bibtexexpression.py]: Could you fix the alignment here please?

File: bibtexparser/bibtexexpression.py [hunk 6]
```diff
@@ -156,10 +168,15 @@ def __init__(self):
...
+        else:
+            self.entry = (entry_type +
+                          in_braces_or_pars(key + pp.Suppress(',') + field_list)
+            )('Entry')
```
Comment (sciunto) [inline bibtexparser/bibtexexpression.py]: and here also.

**Comment by sciunto** [comment]
I put few minor comments. It looks ready to be merged for me. Your unittests are very nice :)

**Comment by sciunto** [comment]
+ you can add your name in the contributors list :)

**Comment by flopraden** [comment]
I'll do the pep8 assigment. Writing the name in contributors list could be you pleasure. But if it's simpler to include it in the PR, I'll do :)

File: bibtexparser/bparser.py (hunk 2 above)
Comment (flopraden) [inline bibtexparser/bparser.py]: Where should I put the doc for this option? I tried to search for doc for common_strings or customization.
tutorial.rst with a custom section?

Comment (sciunto) [inline bibtexparser/bparser.py]: In that file. Look at the top of the line `self.common_strings = common_strings` to see what I mean.

If you would like to add a paragraph to the tutorial, it's definitely welcome :)

**Pull request closed without merging**